block/src/Pos/Block/BlockWorkMode.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE DataKinds #-}
+
+-- | Various constraints needed for block processing.
+
+module Pos.Block.BlockWorkMode
+    ( BlockInstancesConstraint
+    , BlockWorkMode
+    ) where
+
+import           Data.Default (Default)
+import           Ether.Internal (HasLens)
+import           Mockable (Catch, Delay, Fork, Mockables, SharedAtomic, Throw)
+import           System.Wlog (WithLogger)
+
+import           Pos.Binary.Class (Bi)
+import           Pos.Block.Configuration (HasBlockConfiguration)
+import           Pos.Block.Network.Types (MsgBlock, MsgGetBlocks, MsgGetHeaders, MsgHeaders)
+import           Pos.Block.RetrievalQueue (BlockRetrievalQueue, BlockRetrievalQueueTag)
+import           Pos.Block.Slog.Types (HasSlogContext)
+import           Pos.Block.Types (LastKnownHeader, LastKnownHeaderTag, ProgressHeader,
+                                  ProgressHeaderTag, RecoveryHeader, RecoveryHeaderTag)
+import           Pos.Communication.Limits.Types (MessageLimited)
+import           Pos.Communication.Protocol (Message)
+import           Pos.Core.Context (HasPrimaryKey)
+import           Pos.Lrc.Worker (LrcModeFull)
+import           Pos.Recovery.Info (MonadRecoveryInfo)
+import           Pos.Security.Params (SecurityParams)
+import           Pos.Shutdown.Class (HasShutdownContext)
+import           Pos.StateLock (StateLock, StateLockMetrics)
+import           Pos.Txp (GenericTxpLocalData, MempoolExt, MonadTxpLocal, TxpHolderTag)
+import           Pos.Update.Context (UpdateContext)
+import           Pos.Util.TimeWarp (CanJsonLog)
+import           Pos.Util.Util (HasLens')
+
+-- | These instances are implemented in @Pos.Binary.Communication@,
+-- @Pos.Communication.Message@ and @Pos.Communication.Limits@, which
+-- are unavailable at this point, hence we defer providing them
+-- to the calling site.
+type BlockInstancesConstraint =
+    ( Bi MsgGetHeaders
+    , Bi MsgHeaders
+    , Bi MsgGetBlocks
+    , Bi MsgBlock
+    , Message MsgGetHeaders
+    , Message MsgHeaders
+    , Message MsgGetBlocks
+    , Message MsgBlock
+    , MessageLimited MsgGetHeaders
+    , MessageLimited MsgHeaders
+    , MessageLimited MsgGetBlocks
+    , MessageLimited MsgBlock
+    )
```

Comment:
Btw, you could've used `Each` like here https://github.com/input-output-hk/cardano-sl/blob/c60901742e5fab76f34f7207ee91960e2da93131/ssc/Pos/Ssc/Message.hs#L96, but it's 100% minor.